ring/partition_instance_ring.go

```diff
@@ -0,0 +1,123 @@
+package ring
+
+import (
+	"fmt"
+	"time"
+)
+
+type PartitionRingReader interface {
+	PartitionRing() *PartitionRing
+}
+
+// PartitionInstanceRing holds a partitions ring and a instances ring, and provide functions
+// to look up the intersection of the two (e.g. healthy instances by partition).
+type PartitionInstanceRing struct {
+	// partitionsRingReader and partitionsRing are mutually exclusive.
+	partitionsRingReader PartitionRingReader
+	partitionsRing       *PartitionRing
+
+	instancesRing    *Ring
+	heartbeatTimeout time.Duration
+}
+
+func NewPartitionInstanceRing(partitionsRingWatcher PartitionRingReader, instancesRing *Ring, heartbeatTimeout time.Duration) *PartitionInstanceRing {
+	return &PartitionInstanceRing{
+		partitionsRingReader: partitionsRingWatcher,
+		instancesRing:        instancesRing,
+		heartbeatTimeout:     heartbeatTimeout,
+	}
+}
+
+func newStaticPartitionInstanceRing(partitionsRing *PartitionRing, instancesRing *Ring, heartbeatTimeout time.Duration) *PartitionInstanceRing {
+	return &PartitionInstanceRing{
+		partitionsRing:   partitionsRing,
+		instancesRing:    instancesRing,
+		heartbeatTimeout: heartbeatTimeout,
+	}
+}
+
+func (r *PartitionInstanceRing) PartitionRing() *PartitionRing {
+	if r.partitionsRingReader != nil {
+		return r.partitionsRingReader.PartitionRing()
+	}
+
+	return r.partitionsRing
+}
```

Comment:
If you accept my above suggestion about "static" implementation of `PartitionRingReader`, we can get simplify this.

Reply:
Sure. Done in https://github.com/grafana/dskit/pull/479/commits/b457572c7c8cbf10b4e7cce98ae3180920821c91.